Obstacle areas (issue 120)

Davidek523 opened the pull request:
# Pull Request Template

## Type of Pull Request
Only check one box per pull request.
- [x] 🛠️ Code - Changes to the codebase.
- [ ] 🔄 Revert - Reverts a previously merged commit or PR.
- [ ] 📄 Documentation - Updates to documentation files.
- [ ] 🎵 Audio - Changes to audio files.
- [ ] 🎨 Asset - Updates to design or visual assets.

## Summary

<!-- 
INSTRUCTIONS:
- Briefly describe the purpose of this pull request. 
-->

This PR <!-- adds|modifies|updates|fixes|reverts --> adds <!-- Briefly explain what the PR changes. --> obstacles objects to the game. Now when the player comes in contact with such a object, the player will receive damage and on the screen it will be shown how much!


## Related Issues
<!-- 
INSTRUCTIONS:
- Mention related issue numbers here.
- Example: "Fixes #123, Closes #456."
-->

This PR <!-- fixes|closes|addresses|reverts --> closes issue(s) #120   <!-- (e.g., Fixes #123, Closes #456) -->

---

## ✅ Checklist:

- [x] Tested the changes locally.
- [x] Verified that no breaking changes are introduced.
- [ ] Updated documentation (if applicable).


## 🏷️ Labels:
`type: ...`, `area: ...`, `game-...`, ... <!-- Fill and expand on this line as needed. See https://github.com/PyCeas/Pyceas/wiki/Guide-to-using-labels for more information. -->

Davidek523 committed:
Added new class

Added:
- new state for obstacles
- state methods within this class

Davidek523 committed:
Added obstacles

Added:
- obstacles to the map
- in setup drew the obstacles onto the map
- in obstacle state added logic for collision

Davidek523 committed:
fixed some small issues

Fixed:
- in game running:
Typo from the imported class
- obstacle state:
logical issues with collision

Davidek523 committed:
Changing some file structure

Changed:
- Removed the object state from states and refactored it to not use state anymore
- Changed file name and the file place

Davidek523 committed:
Calling the class in game running

Called:
- object handler in game running

Davidek523 committed:
Fixed setup issue

Fixed:
- issue where message of dealing damage was not showing up
Changed:
- commentend display flip method in object handler to save frames

Davidek523 committed:
Making the black screen transparent

Changed:
- the black block for the message transparent

Davidek523 committed:
Changed color

Changed:
- the color of the message

Davidek523 committed:
Merge branch 'main' of https://github.com/PyCeas/Pyceas into Obstacle-areas-(issue-120)

Davidek523 committed:
Small fix

Small fix groups

Davidek523 committed:
Typehints check

Fixed all the typehints

ultimateownsz reviewed (changes requested):
Hi, since there is no objective for the player yet, I think what could work with the collisions is that is the player health < 0, that it will reset the player's position to the starting position.
![image](https://github.com/user-attachments/assets/752c4b1f-705a-4b95-9b4d-14446ab5ba1d)

---

Also, I think it could be helpful if you make it that you see the player getting a white border or flashing sprite that it shows to the player that you hit an object, because now it is static, and the drawing order could be different, so the player's (rectangle for now) behind the obstacle, instead of over the obstacle.
![image](https://github.com/user-attachments/assets/4714d039-97a8-4563-b43c-1b7e18b8815b)

Davidek523 committed:
Fixed health issue

Fixed:
- Health going infinitly low

Davidek523 committed:
Created a flash

Added:
- a flashing effect when the player gets hurt by the obstacles

Davidek523 committed:
Changed the location of the dt

Changed:
- the location of the dt from the handler to the game loop

Davidek523 committed:
Fixed some minor issues with the flash

Fixed:
- Flash not working properly
Removed:
- unnecessary delta time

Davidek523 committed:
Finished flashing feature

Finished:
- when the player comes in contact with a obstacle the rect will flash

Davidek523 committed:
Added respawning

Added:
- when the player reaches 0 health, he will respawn at the start position with the maximum amount of health

Davidek523 committed:
Fixed some typehint issues

Fixed:
- typehint issues

Davidek523 commented:
Addressed the issues that you mentioned! Could you take a look?

ultimateownsz commented:
It's looking good, could you also add the collision with islands?

Davidek523 committed:
Fixed constant flashing

Fixed:
- constant flashing when the player stays on the obstacle.

Davidek523 committed:
Changed file name

Changed:
- the file name to object_manager
- class name to ObjectManager
- as requested

ultimateownsz reviewed (changes requested):
Looking good, it all works. Before I forget, I would only want you to add unit tests since your `obstacle_manager` is a new feature for the project. Do this and your code can be merged!

Davidek523 committed:
Merge branch 'main' into Obstacle-areas-(issue-120)
